src/Compiler/Microsoft.AspNetCore.Razor.Language/src/RazorProjectEngineBuilder.cs

```diff
@@ -1,21 +1,50 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
-using System.Collections.Generic;
+using System;
+using System.Collections.Immutable;
+using Microsoft.AspNetCore.Razor.PooledObjects;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-public abstract class RazorProjectEngineBuilder
+public sealed class RazorProjectEngineBuilder
 {
-    public abstract RazorConfiguration Configuration { get; }
-
-    public abstract RazorProjectFileSystem FileSystem { get; }
-
-    public abstract ICollection<IRazorFeature> Features { get; }
-
-    public abstract IList<IRazorEnginePhase> Phases { get; }
-
-    public abstract RazorProjectEngine Build();
+    public RazorConfiguration Configuration { get; }
+    public RazorProjectFileSystem FileSystem { get; }
+    public ImmutableArray<IRazorFeature>.Builder Features { get; }
+    public ImmutableArray<IRazorEnginePhase>.Builder Phases { get; }
+
+    internal RazorProjectEngineBuilder(RazorConfiguration configuration, RazorProjectFileSystem fileSystem)
+    {
+        Configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+        FileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+        Features = ImmutableArray.CreateBuilder<IRazorFeature>();
+        Phases = ImmutableArray.CreateBuilder<IRazorEnginePhase>();
+    }
+
+    public RazorProjectEngine Build()
+    {
+        using var engineFeatures = new PooledArrayBuilder<IRazorEngineFeature>(Features.Count);
+        using var projectEngineFeatures = new PooledArrayBuilder<IRazorProjectEngineFeature>(Features.Count);
+
+        foreach (var feature in Features)
+        {
+            switch (feature)
+            {
+                case IRazorEngineFeature engineFeature:
+                    engineFeatures.Add(engineFeature);
+                    break;
+
+                case IRazorProjectEngineFeature projectEngineFeature:
+                    projectEngineFeatures.Add(projectEngineFeature);
+                    break;
```

Comment:
Is it worh having a `default:` here that we assert on? Or is this an estalbished pattern?

Reply:
I think it's good to add an assert here. A quick FAR on `IRazorFeature` shows that everything actually goes through `IRazorEngineFeature` and `IRazorProjectEngineFeature`. So, we should probably add a `Debug.Fail()` if some other sort of `IRazorFeature` is passed through here, or if something implements `IRazorFeature` directly.